isis/tests/FunctionalTestsFillGap.cpp

```diff
@@ -0,0 +1,409 @@
+  EXPECT_EQ( (int)dims["Lines"], 9 );
+  EXPECT_EQ( (int)dims["Samples"], 9 );
+  EXPECT_EQ( (int)dims["Bands"], 9 );
+
+  std::unique_ptr<Histogram> outHist (outCube.histogram());
+
+  EXPECT_NEAR( outHist->Average(), 0.058289, 1e-6 );
+  EXPECT_NEAR( outHist->Sum(), 4.721464, 1e-6);
+  EXPECT_EQ( outHist->ValidPixels(), 81 );
+  EXPECT_NEAR( outHist->StandardDeviation(), 1.001199, 1e-6 );
+  EXPECT_NEAR( outHist->Variance(), 1.002400, 1e-6 );
+
+  EXPECT_EQ( outHist->LisPixels(), 0 );
+  EXPECT_EQ( outHist->LrsPixels(), 0 );
+  EXPECT_EQ( outHist->HisPixels(), 0 );
+  EXPECT_EQ( outHist->HrsPixels(), 0 );
+}
+
+TEST_F( SmallGapCube, FillGapTestVertLinear )
+{
+  QTemporaryDir prefix;
+  QString cubeFileName = prefix.path() + "/fillgap_out.cub";
+  QVector<QString> args = {"from=" + vertCube->fileName(),
+                           "to=" + cubeFileName,
+                           "direction=line", "interp=linear" };
+  UserInterface options(APP_XML, args);
+  Pvl log;
+
+  try {
+    fillgap(options, &log);
+  }
+  catch (IException &e) {
+    FAIL() << e.toString().toStdString().c_str() << std::endl;
+  }
+
+  Cube outCube(cubeFileName);
+  Pvl *label = outCube.label();
+
+  PvlGroup &dims = label->findGroup("Dimensions", Pvl::Traverse);
+  EXPECT_EQ( (int)dims["Lines"], 9 );
+  EXPECT_EQ( (int)dims["Samples"], 9 );
+  EXPECT_EQ( (int)dims["Bands"], 9 );
+
+  std::unique_ptr<Histogram> outHist (outCube.histogram());
+
+  EXPECT_NEAR( outHist->Average(), 0.084880, 1e-6 );
+  EXPECT_NEAR( outHist->Sum(), 6.875352, 1e-6);
+  EXPECT_EQ( outHist->ValidPixels(), 81 );
+  EXPECT_NEAR( outHist->StandardDeviation(), 0.975967, 1e-6 );
+  EXPECT_NEAR( outHist->Variance(), 0.952512, 1e-6 );
+
+  EXPECT_EQ( outHist->LisPixels(), 0 );
+  EXPECT_EQ( outHist->LrsPixels(), 0 );
+  EXPECT_EQ( outHist->HisPixels(), 0 );
+  EXPECT_EQ( outHist->HrsPixels(), 0 );
+}
+
+// Testing for logged warning when the app cannot interpolate over
+// special pixels due to a gap taking place on the edge of the cube
+TEST_F( SmallGapCube, FillGapTestGapsOnEdge )
+{
+  QTemporaryDir prefix;
+  QString cubeFileName = prefix.path() + "/fillgap_out.cub";
+  QVector<QString> args = {"from=" + vertCube->fileName(),
+                           "to=" + cubeFileName,
+                           "direction=line" };
+  UserInterface options(APP_XML, args);
+  Pvl log;
+
+
+  // add gap line on edge
+  LineManager line(*vertCube);
+  int lineNum = 0;
+  for(line.begin(); !line.end(); line++) {
+    for(int i = 0; i < line.size(); i++) {
+      if( lineNum == 80 ) { 
+        line[i] = NULL8;
+      }
+    }
+    lineNum++;
+    vertCube->write(line);
+  }
+  vertCube->reopen("rw");
+
+
+  try {
+    fillgap(options, &log);
+  }
+  catch (IException &e) {
+    FAIL() << e.toString().toStdString().c_str() << std::endl;
+  }
+
+  PvlGroup &mess = log.findGroup("Messages", Pvl::Traverse);
+  EXPECT_EQ(mess["Warning"][0].toStdString(), "Unable to fill 9 special pixels." ); 
```

Comment:
Jenkins is still failing on this test but only for Mac. It is saying it expects "Unable to fill 169 special pixels". I find it strange that this is passing on linux though.

Reply:
Also noticed SmallCube.FunctionalTestBandnormByNumber is now segfaulting just on this PR for Mac. Not sure what could cause that in this PR.

Reply:
I think Mac does not like the way I was writing to the cube for this test. Lines 386-399 here were causing some issues before my latest commit. I changed the edge-gap to be on the 0th line instead of the 81st, and changed it to only write to the cube on the 0th line. This solved the Mac failure and the SmallCube segfault.

Reply:
Actually, I'm seeing that SmallCube.FunctionalTestBandnormByNumber segfault in another unrelated PR now. See PR4402 build 7 for Ubuntu.